src/urllib3/util/retry.py

```diff
@@ -179,7 +179,7 @@ def __init__(self, total=10, connect=None, read=None, redirect=None, status=None
         self.raise_on_status = raise_on_status
         self.history = history or tuple()
         self.respect_retry_after_header = respect_retry_after_header
-        self.remove_headers_on_redirect = remove_headers_on_redirect
+        self.remove_headers_on_redirect = [h.lower() for h in remove_headers_on_redirect]
```

Comment:
I think we should change this structure into a frozenset to protect users from adding headers not through the constructor if we're going to enforce lowercase-only.

Reply:
You mean we should remove `remove_headers_on_redirect` parameter from the constructor?

Reply:
No, I mean converting the iterable we receive from remove_headers_on_redirect into a frozenset. Something like `self.remove_headers_on_redirect = frozenset([h.lower() for h in remove_headers_on_redirect])`